src/tiles/storage.rs

```diff
@@ -86,20 +86,41 @@ impl TileStorage {
         self.tiles.iter_mut()
     }
 
-    /// Remove entity at the given tile position, if there was one, leaving `None` in its place.
+    /// Removes any stored `Entity` at the given tile position, leaving `None` in its place and
+    /// returning the `Entity`.
     ///
     /// Panics if the given `tile_pos` doesn't lie within the extents of the underlying tile map.
-    pub fn remove(&mut self, tile_pos: &TilePos) {
-        self.tiles[tile_pos.to_index(&self.size)].take();
+    pub fn remove(&mut self, tile_pos: &TilePos) -> Option<Entity> {
+        self.tiles[tile_pos.to_index(&self.size)].take()
     }
 
-    /// Remove any stored entity at the given tile position, if the given `tile_pos` does lie within
-    /// the extents of the underlying map.
+    /// Remove any stored `Entity` at the given tile position, leaving `None` in its place and
+    /// returning the `Entity`.
     ///
-    /// Otherwise, nothing is done.
-    pub fn checked_remove(&mut self, tile_pos: &TilePos) {
+    /// Checks that the given `tile_pos` lies within the extents of the underlying map.
+    pub fn checked_remove(&mut self, tile_pos: &TilePos) -> Option<Entity> {
         if tile_pos.within_map_bounds(&self.size) {
-            self.tiles[tile_pos.to_index(&self.size)].take();
+            return self.tiles[tile_pos.to_index(&self.size)].take();
         }
+
+        None
```

Comment:
looks fine to me. the index access probably implies that this is guaranteed to be filled with None but just reading the code as an outside observer it could easily look like it was a mistake as well.

I think if we wanted to be exact we'd return `Result<Option<Entity>, BoundsError>` but checking if a tile is "in-bounds" seems trivial for users to do outside of this function if they want that information.